Test to develop - Stellar burgers

Kucsoo opened the pull request.

Kucsoo committed:
Добавлены тесты, локаторы,фикстуры в readme добавлено описание проекта

kadirovgm commented on README.md:
```diff
@@ -1 +1,26 @@
+7) test_constructor_goto_buns_success- проверка перехода в раздел «Булки»
+
+Locators - содержит описания всех локаторов, которые использованы в тестах
+conftest - фикстуры для тестов
```
Отлично! хорошее описание `readme` файла

kadirovgm commented on Tests/Locators.py:
Можно лучше: думаю в папке `Tests` лучше, чтобы были только тесты, а `Locators.py` и `conftest.py` нужно вынести из этой папки

kadirovgm commented on Tests/Locators.py:
```diff
@@ -0,0 +1,25 @@
+from selenium.webdriver.common.by import By
+
+class TestLocators:
```
Можно лучше: класс с локаторами лучше разделить на несколько: например `AuthPageLocators`, `MainPageLocators`
Сейчас это не критично, но в будущем это точно стоит сделать, так как таким образом ты не будешь испортить лишние локаторы в файл с тестами на экран авторизации. Т.е. у каждой страницы должен быть свой класс с локаторами. 

При таком подходе код станет лаконичнее, легко поддерживаемой. Ведь у экрана авторизации и у экрана главной могут быть свои NEXT_BUTTON например. Тогда можно будет легко оба обозвать именно так: 
`MainPageLocators.NEXT_BUTTON` и `AuthPageLocators.NEXT_BUTTON`

kadirovgm commented on Tests/Locators.py:
```diff
@@ -0,0 +1,25 @@
+    HEADER_COLLECT_YOUR_BURGER = ".//h1[contains(test(), 'Соберите бургер')]"
+    LOGO = By.CLASS_NAME, "AppHeader_header__logo__2D0X2"
+    HEADER_SOUCES = By.XPATH, ".//span[contains(text(),'Соусы')]"
+    HEADER_FILLINGS = By.XPATH, ".//span[contains(text(),'Начинки')]"
```
Отлично! хорошее описание локаторов

kadirovgm commented on Tests/Test goto.py:
Нужно исправить: названия для тестовых файлов для pytest должно быть описано по формату:
`test_<something>.py`

kadirovgm commented on Tests/Test login_logout.py:
Нужно исправить: названия для тестовых файлов для pytest должно быть описано по формату:
`test_<something>.py`

kadirovgm commented on Tests/Test registration.py:
Нужно исправить: названия для тестовых файлов для pytest должно быть описано по формату:
`test_<something>.py`

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    driver.get('https://stellarburgers.nomoreparties.site')
+    driver.find_element(*TestLocators.PERSONAL_ACCOUNT).click()
+    WebDriverWait(driver, 5).until(
+        expected_conditions.presence_of_element_located((By.XPATH, ".//button[contains(text(),'Войти')]")))
```
нужно исправить: почему не вынесла в файл с локаторами? в идеале в тестовых файлах локаторов не должно оставаться

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    WebDriverWait(driver, 5).until(
+        expected_conditions.presence_of_element_located((By.XPATH, ".//h2[contains(text(),'Булки')]")))
+    driver.find_element(*TestLocators.PERSONAL_ACCOUNT).click()
+    WebDriverWait(driver, 5).until(expected_conditions.presence_of_element_located(
```
нужно исправить: почему не вынесла в файл с локаторами? в идеале в тестовых файлах локаторов не должно оставаться

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    WebDriverWait(driver, 5).until(
+        expected_conditions.presence_of_element_located((By.XPATH, ".//h2[contains(text(),'Булки')]")))
+    driver.find_element(*TestLocators.PERSONAL_ACCOUNT).click()
+    WebDriverWait(driver, 5).until(expected_conditions.presence_of_element_located(
```
нужно исправить: почему не вынесла в файл с локаторами? в идеале в тестовых файлах локаторов не должно оставаться

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys("kucsoo@ya.ru")
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys("1234567")
+    driver.find_element(*TestLocators.ENTER_BUTTON).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: почему не вынесла в файл с локаторами? в идеале в тестовых файлах локаторов не должно оставаться

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+def test_goto_from_account_to_main_page_by_logo_success():
+    driver.get('https://stellarburgers.nomoreparties.site')
+    driver.find_element(*TestLocators.LOGIN_BUTTON).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys("kucsoo@ya.ru")
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys("1234567")
+    driver.find_element(*TestLocators.ENTER_BUTTON).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys("kucsoo@ya.ru")
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys("1234567")
+    driver.find_element(*TestLocators.ENTER_BUTTON).click()
+    WebDriverWait(driver, 5).until(expected_conditions.presence_of_element_located((By.XPATH, ".//h2[contains(text(),'Булки')]")))
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+def test_login_by_button_enter_to_personal_account_success():
+    driver.get('https://stellarburgers.nomoreparties.site')
+    driver.find_element(*TestLocators.LOGIN_BUTTON).click()
+    WebDriverWait(driver, 5).until(expected_conditions.presence_of_element_located((By.XPATH, ".//button[contains(text(),'Войти')]")))
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+def test_login_by_personal_account_in_header_success():
+    driver.get('https://stellarburgers.nomoreparties.site')
+    driver.find_element(*TestLocators.PERSONAL_ACCOUNT).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+    element = driver.find_element(*TestLocators.ENTER_REGISTRATION_PAGE)
+    driver.execute_script("arguments[0].scrollIntoView();", element)
+    driver.find_element(*TestLocators.ENTER_REGISTRATION_PAGE).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys("kucsoo@ya.ru")
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys("1234567")
+    driver.find_element(*TestLocators.ENTER_BUTTON).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys("kucsoo@ya.ru")
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys("1234567")
+    driver.find_element(*TestLocators.ENTER_BUTTON).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys("1234567")
+    driver.find_element(*TestLocators.ENTER_BUTTON).click()
+    driver.find_element(*TestLocators.PERSONAL_ACCOUNT).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test login_logout.py:
```diff
@@ -0,0 +1,77 @@
+    WebDriverWait(driver, 5).until(
+        expected_conditions.presence_of_element_located((By.XPATH, ".//button[contains(text(),'Выход')]")))
+    driver.find_element(*TestLocators.EXIT_BUTTON).click()
+    WebDriverWait(driver, 5).until(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test registration.py:
```diff
@@ -0,0 +1,65 @@
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys(email)
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys(password)
+    driver.find_element(*TestLocators.REGISTRATION_BUTTON).click()
+    WebDriverWait(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/conftest.py:
```diff
@@ -0,0 +1,44 @@
+import pytest
+import random as r
+
+def email_generation():
```
Можно лучше: это не фикстура, а скорее `helper`, поэтому его лучше не держать в файле `conftest.py`

kadirovgm commented on Tests/conftest.py:
```diff
@@ -0,0 +1,44 @@
+    return email
+
+
+def password_generation():
```
Можно лучше: это не фикстура, а скорее `helper`, поэтому его лучше не держать в файле `conftest.py`

kadirovgm commented on Tests/conftest.py:
```diff
@@ -0,0 +1,44 @@
+    password = str(password_r)
+    return password
+
+def wrong_password_generation():
```
Можно лучше: это не фикстура, а скорее `helper`, поэтому его лучше не держать в файле `conftest.py`

kadirovgm commented on Tests/conftest.py:
```diff
@@ -0,0 +1,44 @@
+    return incorrect_password
+
+
+
```
Нужно исправить: необходимо добавить самую важную фикстуру, отвечающую за `setup` и `teardown`

Пример: 
```
@pytest.fixture()
def driver():
    driver = webdriver.Chrome()
    driver.get(<url main page>)

    yield driver
    driver.quit()
```

Нужно минимизировать повторяющиеся строки кода, в твоих тестах teardown и setup методы описаны для каждого теста, для более лаконичности необходимо их вынести в conftest.py

kadirovgm commented on Tests/Test registration.py:
```diff
@@ -0,0 +1,65 @@
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys(email)
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys(incorrect_password)
+    driver.find_element(*TestLocators.REGISTRATION_BUTTON).click()
+    WebDriverWait(
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test registration.py:
```diff
@@ -0,0 +1,65 @@
+    driver.execute_script("arguments[0].scrollIntoView();", element)
+    WebDriverWait(
+        driver, 5).until(
+        expected_conditions.presence_of_element_located((By.XPATH, ".//p[contains(text(), 'Такой пользователь уже существует')]")))
```
нужно исправить: вынести в файл с локаторами

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+driver.maximize_window()
+
+def test_goto_personal_account_by_header_button_success():
+    driver.get('https://stellarburgers.nomoreparties.site')
```
Нужно исправить: `setup` метод надо вынести, как фикстуру (см. коммент в файле `conftest.py`)
Можно лучше: `https://stellarburgers.nomoreparties.site` - URL главной, лучше такие статичные тестовые данные вынести в отдельный файл типа `urls.py`

Все это нужно для минимизации повторения в коде, ведь таким образом при масштабировании проекта, при изменениях в коде автотесты будет легче поддерживать, потому что если изменится URL главной, упадут все тесты, в таком случае тебе легко будет поменять URL в файле `urls.py`, вместо того, чтобы в каждом тесте в ручную менять такие данные.

PS. актуально для всех твоих тестов

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    WebDriverWait(driver, 5).until(
+        expected_conditions.presence_of_element_located((By.XPATH, ".//button[contains(text(),'Войти')]")))
+    current_url = driver.current_url
+    assert current_url == 'https://stellarburgers.nomoreparties.site/login', "Переход в личный кабинет не выполнен"
```
Можно лучше: 
- надо вынести статичный url в отдельный файл для переиспользования в коде
- проверка того, что URL поменялся на <url>/login не гарантирует, что страница личного кабинета прогрузилась, поэтому стоит ориентироваться на какие-либо элементы страницы, чекать, что они появились и отображаются корректно

PS. относится ко многим твоим тестам ниже

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+        expected_conditions.presence_of_element_located((By.XPATH, ".//button[contains(text(),'Войти')]")))
+    current_url = driver.current_url
+    assert current_url == 'https://stellarburgers.nomoreparties.site/login', "Переход в личный кабинет не выполнен"
+    driver.quit()
```
Нужно исправить: `setup` метод надо вынести, как фикстуру (см. коммент в файле `conftest.py`)

PS. относится ко всем твоим тестам ниже

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    driver.find_element(*TestLocators.INPUT_PASSWORD).send_keys("1234567")
+    driver.find_element(*TestLocators.ENTER_BUTTON).click()
+    WebDriverWait(driver, 5).until(
+        expected_conditions.presence_of_element_located((By.XPATH, ".//h2[contains(text(),'Булки')]")))
```
нужно исправить: почему не вынесла в файл с локаторами? в идеале в тестовых файлах локаторов не должно оставаться

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+
+def test_goto_profile_by_personal_account_button():
+    driver.get('https://stellarburgers.nomoreparties.site/login')
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys("kucsoo@ya.ru")
```
Можно лучше: статичные тестовые данные лучше выносить в отдельный файл типа data.py - таким образом избегаем повторений в коде и улучшаем поддерживаемость в коде

PS. относится ко всем твоим тестам ниже

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    WebDriverWait(driver, 5).until(expected_conditions.presence_of_element_located(
+        (By.XPATH, ".//button[contains(test(), 'Сохранить')]")))
+    current_url = driver.current_url
+    assert current_url == 'https://stellarburgers.nomoreparties.site/account/profile', "Переход в личный кабинет не выполнен"
```
Нужно исправить: здесь также, проверка URL страницы не гарантирует, что страница прогрузилась и отображается корректно

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+
+def test_goto_constructor_from_personal_account_success():
+    driver.get('https://stellarburgers.nomoreparties.site/login')
+    driver.find_element(*TestLocators.INPUT_EMAIL).send_keys("kucsoo@ya.ru")
```
Можно лучше: видишь, ты часто используешь эти строки в тестах, в таком случае стоит задуматься о вынесении таких строк в отдельную фикстуру, или просто как отдлеьную функцию, помогающая зарегистрироваться

kadirovgm commented on Tests/Test goto.py:
```diff
@@ -0,0 +1,77 @@
+    WebDriverWait(driver, 5).until(expected_conditions.presence_of_element_located(
+        (By.XPATH, ".//button[contains(test(), 'Сохранить')]")))
+    driver.find_element(*TestLocators.CONSTRUCTOR).click()
+    assert driver.find_element(*TestLocators.HEADER_COLLECT_YOUR_BURGER).text == 'Соберите бургер', "Переход в конструктор не выполнен"
```
Отлично! вот здесь корректная проверка, молодец!

Kucsoo committed:
1. Исправлены названия файлов с тестами
2. Локаторы вынесены в отдельный файл
3. Добавлена фикстура setup и teardown
4. Добавлен файл urls со страницами
5. Добавлен файл data с тестовыми данными
6. Добавлен файл helpers, в который венесены фукции из файла conftest
7. В папке Tests остались только файлы с тестами, остальное в корне проекта
8. ОР в тестах исправлен на появление определенных элементов на странице, сравнение url - убрано

kadirovgm commented on conftest.py:
```diff
@@ -0,0 +1,39 @@
+    driver.quit()
+
+
+
```
Молодец! теперь все четко!

kadirovgm commented on helpers.py:
```diff
@@ -0,0 +1,19 @@
+import random as r
+class Helpers:
```
Отлично!
